### ISSUE TITLE
[진행중] conv2d model setting

### ISSUE THREAD

**Issue by skkulsc** [skkulsc/text#2]
Classification_Conv2D Model setting --- 2018-11-29 15:48:27.441838

for_testing = [ False ]

trainingDataPath = [ ../data/merged_dataset/train.txt ]

validationDataPath = [ ../data/merged_dataset/valid.txt ]

testingDataPath = [ ../data/merged_dataset/test.txt ]

wordDicPath = [ ../data/word_embedding_100000/wordDictionary_100000.pkl ] - vocab_size = [ 100000]

Model Training Epochs = [   300]

batch_size = [ 256] - update_size = [ 512] - training_step = [   2]

sequence_length = [ 300] - category_size = [ 6] - embedding_size = [ 512], embedding_trainable = [True]

train_X.shape = [ (520601, 300) ] - valid_X.shape = [ (142510, 300) ]

embeddingWeight.shape = [ (100000, 512) ]

filter_sizes = [3, 4, 5] - num_filters = [ 256]

lr = [ 0.00100] - lr_decay = [ 0.97000] - dropout_keep_prob = [0.50]

train_X.shape = [ (520601, 300) ] - valid_X.shape = [ (142510, 300) ]

save_folder = [ ./Classification_Conv2D/2018-11-29__15:45:46 ]

save_check_point_name = [ ./Classification_Conv2D/2018-11-29__15:45:46/model/Classification_Conv2D_model ]

check_point_file = [ None ]

-------------------------------------------------------------------------------------------

 ********************************** Classification_Conv2D_model structure **********************************

< sequence_inputs > : 
Tensor("sequence_inputs:0", shape=(?, 300), dtype=int32)

< category_inputs > : 
Tensor("category_inputs:0", shape=(?,), dtype=int32)

< one_hot_category_inputs > : 
Tensor("one_hot:0", shape=(?, 6), dtype=float32)

< embeddings > : 
<tf.Variable 'embeddings:0' shape=(100000, 512) dtype=float32_ref>

< sequence_emb > : 
Tensor("sequence_emb/Identity:0", shape=(?, 300, 512), dtype=float32, device=/device:CPU:0)

< sequence_emb_expanded > : 
Tensor("sequence_emb_expanded:0", shape=(?, 300, 512, 1), dtype=float32)

< num_filters_total > : [ 768]

###  filter_size =  3 

< W_conv > : 
<tf.Variable 'conv-maxpool-3/W_conv:0' shape=(3, 512, 1, 256) dtype=float32_ref>
< b_conv > : 
<tf.Variable 'conv-maxpool-3/b_conv:0' shape=(256,) dtype=float32_ref>

< conv > : 
Tensor("conv-maxpool-3/conv:0", shape=(?, 298, 1, 256), dtype=float32)

< h > : 
Tensor("conv-maxpool-3/relu:0", shape=(?, 298, 1, 256), dtype=float32)

### filter_size =  4

< W_conv > : 
<tf.Variable 'conv-maxpool-4/W_conv:0' shape=(4, 512, 1, 256) dtype=float32_ref>

< b_conv > : 
<tf.Variable 'conv-maxpool-4/b_conv:0' shape=(256,) dtype=float32_ref>

< conv > : 
Tensor("conv-maxpool-4/conv:0", shape=(?, 297, 1, 256), dtype=float32)

< h > : 
Tensor("conv-maxpool-4/relu:0", shape=(?, 297, 1, 256), dtype=float32)

### filter_size =  5 

< W_conv > : 
<tf.Variable 'conv-maxpool-5/W_conv:0' shape=(5, 512, 1, 256) dtype=float32_ref>

< b_conv > : 
<tf.Variable 'conv-maxpool-5/b_conv:0' shape=(256,) dtype=float32_ref>

< conv > : 
Tensor("conv-maxpool-5/conv:0", shape=(?, 296, 1, 256), dtype=float32)

< h > : 
Tensor("conv-maxpool-5/relu:0", shape=(?, 296, 1, 256), dtype=float32)

< h_pool > : 
Tensor("conv_flat/h_pool:0", shape=(?, 1, 1, 768), dtype=float32)

< h_pool_flat > : 
Tensor("conv_flat/h_pool_flat:0", shape=(?, 768), dtype=float32)

< h_drop > : 
Tensor("dropout/dropout/mul:0", shape=(?, 768), dtype=float32)

< W_output > : 
<tf.Variable 'output/W_output:0' shape=(768, 6) dtype=float32_ref>

< b_output > : 
<tf.Variable 'output/b_output:0' shape=(6,) dtype=float32_ref>

< scores > : 
Tensor("output/scores:0", shape=(?, 6), dtype=float32)

< loss > : 
Tensor("loss/Mean:0", shape=(), dtype=float32)

< predictions > : 
Tensor("output/predictions:0", shape=(?,), dtype=int64)

< correct_predictions > : 
Tensor("accuracy/Equal:0", shape=(?,), dtype=bool)

< accuracy > : 
Tensor("accuracy/accuracy:0", shape=(), dtype=float32)

< loss_summary > : 
Tensor("summary/train_loss:0", shape=(), dtype=string)

< acc_summary > : 
Tensor("summary/train_accuracy:0", shape=(), dtype=string)

< lr_summary > : 
Tensor("summary/learning_rate:0", shape=(), dtype=string)

< train_summary_op > : 
Tensor("summary/Merge/MergeSummary:0", shape=(), dtype=string)

< test_summary_op > : 
Tensor("summary/Merge_1/MergeSummary:0", shape=(), dtype=string)

< optimizer > : 
<tensorflow.python.training.adam.AdamOptimizer object at 0x7fa0632ce160>

< train_step > : 
name: "optimizer/Adam"
op: "NoOp"
input: "^optimizer/Adam/update_embeddings/ApplyAdam"
input: "^optimizer/Adam/update_conv-maxpool-3/W_conv/ApplyAdam"
input: "^optimizer/Adam/update_conv-maxpool-3/b_conv/ApplyAdam"
input: "^optimizer/Adam/update_conv-maxpool-4/W_conv/ApplyAdam"
input: "^optimizer/Adam/update_conv-maxpool-4/b_conv/ApplyAdam"
input: "^optimizer/Adam/update_conv-maxpool-5/W_conv/ApplyAdam"
input: "^optimizer/Adam/update_conv-maxpool-5/b_conv/ApplyAdam"
input: "^optimizer/Adam/update_output/W_output/ApplyAdam"
input: "^optimizer/Adam/update_output/b_output/ApplyAdam"
input: "^optimizer/Adam/Assign"
input: "^optimizer/Adam/Assign_1"

# train accuracy, loss summary (1 epoch 당 약 2000개의 batch)

<img width="684" alt="train_summary" src="https://user-images.githubusercontent.com/37528988/49237341-591e2700-f441-11e8-85e2-b14d68f01816.png">

# test accuracy, loss summary  (1 epoch당 약 550개의 batch)

<img width="685" alt="test_summary" src="https://user-images.githubusercontent.com/37528988/49237753-3dffe700-f442-11e8-845a-c8e5602a4db0.png">

# graph

## 전체적인 구조

<img width="1022" alt="graph1" src="https://user-images.githubusercontent.com/37528988/49238000-d9915780-f442-11e8-99ea-55aa9d806753.png">

## optimizer

<img width="1569" alt="optimizer" src="https://user-images.githubusercontent.com/37528988/49238009-ddbd7500-f442-11e8-9870-0b6d6777dcae.png">

batch size가 512가 될 시 메모리가 부족한 모델들은 batch size를 128, 256 등으로 낮춰서 전달하며 gradients 값들을 저장한 후, 512개의 data가 통과하면 gradients를 update함. 
예를 들어, batch size가 128이면 4개의 batch에 대한 gradients 값들을 누적한 후, 평균 gradients 값을 구하고model을 update함.

마지막에는 clip_by_global_norm을 적용해 gradients를 조절함.
